### ISSUE TITLE
cargo test fails on recent serde

### ISSUE THREAD

**Issue by FauxFaux** [3Hren/msgpack-rust#178]
With serde newer than 1.0.38, the tests fail:

```
---- round_trip_untagged_enum_with_enum_associated_data stdout ----
thread 'round_trip_untagged_enum_with_enum_associated_data' panicked at 'called `Result::unwrap()` on an `Err` value: Syntax("data did not match any variant of untagged enum Foo")', src/libcore/result.rs:1009:5
```

The test failure can be seen by `cargo update`ing, which will bring you to serde 1.0.80 (or higher)!

To go back in time, try:
```
serde = "=1.0.38"
serde_derive = "=1.0.38"
```

1.0.39 has a different error to the one we see today.

**Comment by daboross** [comment]
I think the culprit for this is https://github.com/serde-rs/serde/commit/e1db820c9fa614eceb23baa4e7f2cbd4fc2dfa21#diff-751ea35094a96d89c2d04ff6a61e703aR2139

Specifically, the addition of `deserialize_identifier` rejecting non-String non-byte things.

Background:

When doing untagged deserialization, the data is all cached into a private serde struct `Content`. `Content` uses no deserialization hints, and just grabs all the data. Then it tries to deserialize each variant of the untagged enum till it finds one that matches - feeding each section with `ContentRefDeserializer`.

The problem is two-fold:

1. When an enum wants to deserialize its variant, it uses the `deserialize_identifier` request. 
2. `ContentRefDeserializer`, as of the above commit, rejects all integers when `deserialize_identifier` is requested.

The enum's variant's `Visitor` implementation would fully accept a `u32` or other integer, and handle it correctly, but because it _requested an identifier_, `ContentRefDeserializer` rejects it immediately.

As for a solution, I'm not sure what the best would be. I think this means that `#[serde(untagged)]` is broken for _all_ formats that encode enum variants as `u32`, though, not just `msgpack`.

**Comment by dtolnay** [comment]
That looks like a Serde bug -- that `deserialize_identifier` should allow u32 to account for formats that encode enum variant that way. Could one of you send a PR to allow u32 there?

**Comment by daboross** [comment]
Will do @dtolnay

**Comment by daboross** [comment]
Submitted https://github.com/serde-rs/serde/pull/1438.

**Comment by daboross** [comment]
https://github.com/serde-rs/serde/pull/1438 has been merged and released as serde `1.0.81`.

I've tested it locally and this seems to be fixed with that serde release!

**Comment by joshtriplett** [comment]
This should be fixed now.
